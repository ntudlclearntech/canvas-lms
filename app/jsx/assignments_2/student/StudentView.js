/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import StudentContainer from './components/StudentContainer'
import {string} from 'prop-types'
import {Query} from 'react-apollo'
import gql from 'graphql-tag'

// Fields I'm not sure if we are using:
//   - state
//   - course
//   - various lids
export const STUDENT_VIEW_QUERY = gql`
  query GetAssignment($assignmentLid: ID!) {
    assignment: legacyNode(type: Assignment, _id: $assignmentLid) {
      ... on Assignment {
        lid: _id
        gid: id
        name
        description
        dueAt
        pointsPossible
        lockAt
        unlockAt
        state
        course {
          lid: _id
        }
        modules {
          name
          lid: _id
        }
        assignmentGroup {
          name
          lid: _id
        }
        lockInfo {
          isLocked
        }
      }
    }
  }
`

const StudentView = props => (
  <Query query={STUDENT_VIEW_QUERY} variables={{assignmentLid: props.assignmentLid}}>
    {({loading, error, data}) => {
      // TODO HANDLE ERROR AND LOADING
      if (loading) return null
      if (error) return `Error!: ${error}`
      return <StudentContainer assignment={data.assignment} />
    }}
  </Query>
)

StudentView.propTypes = {
  assignmentLid: string.isRequired
}

export default React.memo(StudentView)
