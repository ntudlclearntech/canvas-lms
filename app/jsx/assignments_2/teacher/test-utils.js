/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {TeacherViewContextDefaults} from './components/TeacherViewContext'
import {wait} from 'react-testing-library'

// because our version of jsdom doesn't support elt.closest('a') yet. Should soon.
export function closest(el, selector) {
  while (el && !el.matches(selector)) {
    el = el.parentElement
  }
  return el
}

export function findInputForLabel(labelChild, container) {
  const label = closest(labelChild, 'label')
  const input = container.querySelector(`#${label.getAttribute('for')}`)
  return input
}

export function waitForNoElement(queryFn) {
  // use wait instead of waitForElement because waitForElement doesn't seem to
  // trigger the callback when elements disappear
  return wait(() => {
    let elt = null
    try {
      elt = queryFn()
    } catch (e) {
      // if queryFn throws, assume element can't be found and succeed
      return
    }

    // fail if the element was found
    if (elt !== null) throw new Error(`element is still present`)
    // otherwise success
  })
}

export function mockCourse(overrides) {
  return {
    lid: 'course-lid',
    ...overrides
  }
}

export function mockAssignment(overrides) {
  return {
    lid: 'assignment-lid',
    name: 'assignment name',
    pointsPossible: 5,
    dueAt: '2018-11-27T13:00-05:00',
    lockAt: '2018-11-27T13:00-05:00',
    unlockAt: '2018-11-27T13:00-05:00',
    description: 'assignment description',
    state: 'published',
    needsGradingCount: 0,
    course: mockCourse(),
    modules: [{lid: '1', name: 'module 1'}, {lid: '2', name: 'module 2'}],
    assignmentGroup: {lid: '1', name: 'assignment group'},
    lockInfo: {
      isLocked: false
    },
    submissionTypes: ['online_text_entry'],
    allowedExtensions: [],
    assignmentOverrides: {
      nodes: []
    },
    submissions: {
      nodes: []
    },
    ...overrides
  }
}

export function mockOverride(overrides) {
  return {
    gid: '1',
    lid: '1',
    title: 'Section A',
    dueAt: '2018-12-25T23:59:59-05:00',
    allDay: true,
    lockAt: '2018-12-29T23:59:00-05:00',
    unlockAt: '2018-12-23T00:00:00-05:00',
    ...overrides
  }
}

export function mockSubmission(overrides) {
  return {
    gid: '1',
    lid: '1',
    state: 'submitted',
    submissionStatus: 'submitted',
    gradingStatus: 'needs_grading',
    excused: false,
    latePolicyStatus: null,
    submittedAt: '2019-01-17T12:21:42Z',
    user: mockUser(),
    assetId: '2',
    assetType: 'Submission',
    assessorId: '3',
    assessorAssetId: '33',
    assessorAssetType: 'Submission',
    ...overrides
  }
}

export function mockUser(overrides) {
  return {
    lid: 'user_1',
    gid: 'user_1',
    name: 'User 1',
    ...overrides
  }
}

// values need to match the defaults for TeacherViewContext
export const mockTeacherContext = () => ({...TeacherViewContextDefaults})
