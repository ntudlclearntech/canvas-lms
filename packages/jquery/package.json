{
  "name": "jquery",
  "private": true,
  "version": "1.7.2",
  "author": "neme",
  "main": "./jquery.js"
}
