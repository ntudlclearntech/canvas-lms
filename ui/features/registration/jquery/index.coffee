## referenced as extension in gems/plugins
import index from './index.js'
export default index
