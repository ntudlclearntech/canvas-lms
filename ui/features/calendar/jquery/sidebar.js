/*
 * Copyright (C) 2012 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import $ from 'jquery'
import _ from 'underscore'
import React from 'react'
import ReactDOM from 'react-dom'
import {useScope as useI18nScope} from '@canvas/i18n'
import ColorPicker from '@canvas/color-picker'
import userSettings from '@canvas/user-settings'
import contextListTemplate from '../jst/contextList.handlebars'
import forceScreenreaderToReparse from 'force-screenreader-to-reparse'
import 'jquery-kyle-menu'
import '@canvas/jquery/jquery.instructure_misc_helpers'
import 'jquery-tinypubsub'
import 'jqueryui/tooltip'

const I18n = useI18nScope('calendar_sidebar')

class VisibleContextManager {
  constructor(contexts, selectedContexts, $holder) {
    this.$holder = $holder
    const fragmentData = (() => {
      try {
        return $.parseJSON($.decodeFromHex(window.location.hash.substring(1))) || {}
      } catch (e) {
        return {}
      }
    })()

    if (!this.disabledContexts) {
      const disabledContexts = contexts
        .filter(c => c.course_pacing_enabled && !(c.can_make_reservation || c.user_is_observer))
        .map(dC => dC.asset_string)
      this.disabledContexts = disabledContexts
    }

    const availableContexts = contexts
      .filter(ctx => !this.disabledContexts?.includes(ctx.asset_string))
      .map(c => c.asset_string)

    if (fragmentData.show) {
      this.contexts = fragmentData.show.split(',')
    }
    if (!this.contexts) {
      this.contexts = selectedContexts
    }
    if (!this.contexts) {
      this.contexts = availableContexts
    }

    this.contexts = _.intersection(this.contexts, availableContexts)
    this.contexts = this.contexts.slice(0, ENV.CALENDAR.VISIBLE_CONTEXTS_LIMIT)

    this.notify()
    this.disableContexts()

    $.subscribe('Calendar/saveVisibleContextListAndClear', this.saveAndClear)
    $.subscribe('Calendar/restoreVisibleContextList', this.restoreList)
    $.subscribe('Calendar/ensureCourseVisible', this.ensureCourseVisible)
  }

  saveAndClear = () => {
    if (!this.savedContexts) {
      this.savedContexts = this.contexts
      this.contexts = []
      return this.notifyOnChange()
    }
  }

  restoreList = () => {
    if (this.savedContexts) {
      this.contexts = this.savedContexts
      this.savedContexts = null
      return this.notifyOnChange()
    }
  }

  ensureCourseVisible = context => {
    if ($.inArray(context, this.contexts) < 0) {
      return this.toggle(context)
    }
  }

  toggle(context) {
    const index = $.inArray(context, this.contexts)
    if (index >= 0) {
      this.contexts.splice(index, 1)
    } else {
      this.contexts.push(context)
      if (this.contexts.length > ENV.CALENDAR.VISIBLE_CONTEXTS_LIMIT) {
        this.contexts.shift()
      }
    }
    return this.notifyOnChange()
  }

  notifyOnChange = () => {
    this.notify()

    return $.ajaxJSON('/api/v1/calendar_events/save_selected_contexts', 'POST', {
      selected_contexts: this.contexts
    })
  }

  notify = () => {
    $.publish('Calendar/visibleContextListChanged', [this.contexts])

    this.$holder.find('.context_list_context').each((i, li) => {
      let needle
      const $li = $(li)
      const visible = ((needle = $li.data('context')), this.contexts.includes(needle))
      $li
        .toggleClass('checked', visible)
        .toggleClass('not-checked', !visible)
        .find('.context-list-toggle-box')
        .attr('aria-checked', visible)
    })

    return userSettings.set('checked_calendar_codes', this.contexts)
  }

  disableContexts = () => {
    this.$holder.find('.context_list_context').each((i, li) => {
      const $li = $(li)
      const disabled = this.disabledContexts.includes($li.data('context'))
      if (disabled) {
        const label = document.createElement('label')
        label.style.display = 'block'
        label.innerHTML = I18n.t('Due dates managed by Course Pacing.')
        $li[0].appendChild(label)
        $li.toggleClass('disabled-context', disabled)
        $li.attr('title', I18n.t('Course calendar view disabled.'))
        $li.on('click keyclick', false)
        $li.children().attr('tabindex', -1)
        $($li).tooltip({
          position: {my: 'left bottom', at: 'left top'},
          tooltipClass: 'center bottom vertical'
        })
      }
    })
  }
}

function setupCalendarFeedsWithSpecialAccessibilityConsiderationsForNVDA() {
  const $calendarFeedModalContent = $('#calendar_feed_box')
  const $calendarFeedModalOpener = $('.dialog_opener[aria-controls="calendar_feed_box"]')
  // We need to get the modal initialized early rather than wait for
  // .dialog_opener to open it so we can attach the event to it the first
  // time.  We extend so that we still get all the magic that .dialog_opener
  // should give us.
  $calendarFeedModalContent.dialog(
    $.extend(
      {
        autoOpen: false,
        modal: true
      },
      $calendarFeedModalOpener.data('dialogOpts')
    )
  )

  $calendarFeedModalContent.on('dialogclose', () => {
    forceScreenreaderToReparse($('#application')[0])
    $('#calendar-feed .dialog_opener').focus()
  })
}

export default function sidebar(contexts, selectedContexts, dataSource) {
  const $holder = $('#context-list-holder')
  const $skipLink = $('.skip-to-calendar')
  const $colorPickerBtn = $('.ContextList__MoreBtn')

  $skipLink.focus(() => {
    $skipLink.removeClass('screenreader-only')
  })

  $skipLink.focusout(() => {
    $skipLink.addClass('screenreader-only')
  })

  setupCalendarFeedsWithSpecialAccessibilityConsiderationsForNVDA()

  $holder.html(contextListTemplate({contexts}))

  const visibleContexts = new VisibleContextManager(contexts, selectedContexts, $holder)

  $holder.on('click keyclick', '.context-list-toggle-box', function (event) {
    const parent = $(this).closest('.context_list_context')
    visibleContexts.toggle($(parent).data('context'))
  })

  $holder.on('click keyclick', '.ContextList__MoreBtn', function (event) {
    const positions = {
      top: $(this).offset().top - $(window).scrollTop(),
      left: $(this).offset().left - $(window).scrollLeft()
    }

    const assetString = $(this).closest('li').data('context')

    // ensures previously picked color clears
    ReactDOM.unmountComponentAtNode($('#color_picker_holder')[0])

    ReactDOM.render(
      <ColorPicker
        isOpen
        positions={positions}
        assetString={assetString}
        afterClose={() => forceScreenreaderToReparse($('#application')[0])}
        afterUpdateColor={color => {
          color = `#${color}`
          const $existingStyles = $('#calendar_color_style_overrides')
          const $newStyles = $('<style>')
          $newStyles.text(
            `.group_${assetString},
            .group_${assetString}:hover,
            .group_${assetString}:focus{
              color: ${color};
              border-color: ${color};
              background-color: ${color};
            }`
          )
          $existingStyles.append($newStyles)
        }}
      />,
      $('#color_picker_holder')[0]
    )
  })

  $skipLink.on('click', e => {
    e.preventDefault()
    $('#content').attr('tabindex', -1).focus()
  })
}
