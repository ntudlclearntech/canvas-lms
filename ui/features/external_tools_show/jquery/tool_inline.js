/*
 * Copyright (C) 2011 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import $ from 'jquery'
import '@canvas/module-sequence-footer'
import MarkAsDone from '@canvas/util/jquery/markAsDone'
import ToolLaunchResizer from '@canvas/lti/jquery/tool_launch_resizer'
import {monitorLtiMessages} from '@canvas/lti/jquery/messages'
import ready from '@instructure/ready'

ready(() => {
  const $toolForm = $('#tool_form')

  const launchToolManually = function () {
    const $button = $toolForm.find('button')

    $toolForm.show()

    // Firefox remembers disabled state after page reloads
    $button.attr('disabled', false)
    setTimeout(() => {
      // LTI links have a time component in the signature and will
      // expire after a few minutes.
      $button.attr('disabled', true).text($button.data('expired_message'))
    }, 60 * 2.5 * 1000)

    $toolForm.submit(function () {
      $(this).find('.load_tab,.tab_loaded').toggle()
    })
  }

  const launchToolInNewTab = function () {
    $toolForm.attr('target', '_blank')
    launchToolManually()
  }

  switch ($toolForm.data('tool-launch-type')) {
    case 'window':
      $toolForm.show()
      launchToolInNewTab()
      break
    case 'self':
      $toolForm.removeAttr('target')
      try {
        $toolForm.submit()
        // eslint-disable-next-line no-empty
      } catch (e) {}
      break
    default:
      // Firefox throws an error when submitting insecure content
      try {
        $toolForm.submit()
        // eslint-disable-next-line no-empty
      } catch (e) {}

      $('#tool_content').bind('load', () => {
        if (
          document.location.protocol !== 'https:' ||
          $('#tool_form')[0].action.indexOf('https:') > -1
        ) {
          $('#insecure_content_msg').hide()
          $toolForm.hide()
        }
      })
      setTimeout(() => {
        if ($('#insecure_content_msg').is(':visible')) {
          $('#load_failure').show()
          launchToolInNewTab()
        }
      }, 3 * 1000)
      break
  }

  // Iframe resize handler
  const $tool_content_wrapper = $('.tool_content_wrapper')
  let tool_height, canvas_chrome_height

  const $window = $(window)
  const toolResizer = new ToolLaunchResizer(tool_height)

  const $external_content_info_alerts = $tool_content_wrapper.find(
    '.before_external_content_info_alert, .after_external_content_info_alert'
  )

  $external_content_info_alerts.on('focus', function () {
    $tool_content_wrapper.find('iframe').css('border', '2px solid #0374B5')
    $(this).removeClass('screenreader-only-tool')
  })

  $external_content_info_alerts.on('blur', function () {
    $tool_content_wrapper.find('iframe').css('border', 'none')
    $(this).addClass('screenreader-only-tool')
  })

  const is_full_screen = $('body').hasClass('ic-full-screen-lti-tool')

  if (!is_full_screen) {
    canvas_chrome_height = $tool_content_wrapper.offset().top + $('#footer').outerHeight(true)
  }

  if ($tool_content_wrapper.length) {
    $window
      .on('resize', () => {
        // https://api.jquery.com/resize/
        // https://developer.mozilla.org/en-US/docs/Web/API/Window/resize_event

        if (!$tool_content_wrapper.data('height_overridden')) {
          if (is_full_screen) {
            // divs from app/views/lti/_lti_message.html.erb that usually have 1px
            const div_before_iframe =
              document.querySelector('div.before_external_content_info_alert')?.offsetHeight || 0
            const div_after_iframe =
              document.querySelector('div.after_external_content_info_alert')?.offsetHeight || 0

            // header#mobile-header
            //   hidden when screen width > 768px
            //   see app/stylesheets/base/_ic_app_header.scss
            //   see https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetHeight
            const mobile_header_height =
              document.querySelector('header#mobile-header')?.offsetHeight || 0

            tool_height =
              window.innerHeight - mobile_header_height - div_before_iframe - div_after_iframe

            toolResizer.resize_tool_content_wrapper(tool_height, $tool_content_wrapper, true)
          } else {
            toolResizer.resize_tool_content_wrapper(
              $window.height() -
                canvas_chrome_height -
                // module item navigation from PLAT-1687
                $('#sequence_footer').outerHeight(true)
            )
          }
        }
      })
      .triggerHandler('resize')
  }

  if (ENV.LTI != null && ENV.LTI.SEQUENCE != null) {
    $('#module_sequence_footer').moduleSequenceFooter({
      assetType: 'Lti',
      assetID: ENV.LTI.SEQUENCE.ASSET_ID,
      courseID: ENV.LTI.SEQUENCE.COURSE_ID
    })
  }

  $('#content').on('click', '#mark-as-done-checkbox', function () {
    MarkAsDone.toggle(this)
  })
})

monitorLtiMessages()
