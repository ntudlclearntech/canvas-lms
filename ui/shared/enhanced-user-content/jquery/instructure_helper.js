/*
 * Copyright (C) 2011 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

// create a global object "INST" that we will have be Instructure's namespace.
import $ from 'jquery'
import 'jqueryui/dialog'

window.equella = {
  ready(data) {
    $(document).triggerHandler('equella_ready', data)
  },
  cancel() {
    $(document).triggerHandler('equella_cancel')
  },
}
$(document)
  .bind('equella_ready', function (event, data) {
    $('#equella_dialog').triggerHandler('equella_ready', data)
  })
  .bind('equella_cancel', function () {
    $('#equella_dialog').dialog('close')
  })

window.external_tool_dialog = {
  ready(data) {
    const e = $.Event('selection')
    e.contentItems = data
    $('#resource_selection_dialog:visible').triggerHandler(e)
    $('#homework_selection_dialog:visible').triggerHandler(e)
  },
  cancel() {
    $('#external_tool_button_dialog').dialog('close')
    $('#resource_selection_dialog').dialog('close')
    $('#homework_selection_dialog:visible').dialog('close')
  },
}
