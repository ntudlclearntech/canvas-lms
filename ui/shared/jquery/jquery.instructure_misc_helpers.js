/*
 * Copyright (C) 2011 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import $ from 'jquery'
import htmlEscape from 'html-escape'

if (!('INST' in window)) window.INST = {}

// useful for i18n, e.g. t('key', 'pick one: %{select}', {select: $.raw('<select><option>...')})
// note that raw returns a SafeString object, so you may want to call toString
// if you're using it elsewhere
$.raw = function (str) {
  return new htmlEscape.SafeString(str)
}
// ensure the jquery html setters don't puke if given a SafeString
$.each(['html', 'append', 'prepend'], function (idx, method) {
  const orig = $.fn[method]
  $.fn[method] = function () {
    const args = [].slice.call(arguments)
    for (let i = 0, len = args.length; i < len; i++) {
      if (args[i] instanceof htmlEscape.SafeString) args[i] = args[i].toString()
    }
    return orig.apply(this, args)
  }
})

$.replaceOneTag = function (text, name, value) {
  if (!text) {
    return text
  }
  name = (name || '').toString()
  value = (value || '').toString().replace(/\s/g, '+')
  const itemExpression = new RegExp('(%7B|{){2}[\\s|%20|+]*' + name + '[\\s|%20|+]*(%7D|}){2}', 'g')
  return text.replace(itemExpression, value)
}
// backwards compatible with only one tag
$.replaceTags = function (text, mapping_or_name, maybe_value) {
  if (typeof mapping_or_name === 'object') {
    for (const name in mapping_or_name) {
      text = $.replaceOneTag(text, name, mapping_or_name[name])
    }
    return text
  } else {
    return $.replaceOneTag(text, mapping_or_name, maybe_value)
  }
}

// return query string parameter
// $.queryParam("name") => qs value or null
$.queryParam = function (name) {
  name = name.replace(/[\[]/, '\\[').replace(/[\]]/, '\\]')
  const regex = new RegExp('[\\?&]' + name + '=([^&#]*)')
  const results = regex.exec(window.location.search)
  if (results == null) return results
  else return decodeURIComponent(results[1].replace(/\+/g, ' '))
}