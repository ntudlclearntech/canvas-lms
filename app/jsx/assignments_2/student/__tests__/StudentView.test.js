/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import ReactDOM from 'react-dom'
import $ from 'jquery'
import {MockedProvider} from 'react-apollo/test-utils'
import StudentView, {STUDENT_VIEW_QUERY} from '../StudentView'
import wait from 'waait'

const mocks = [
  {
    request: {
      query: STUDENT_VIEW_QUERY,
      variables: {
        assignmentLid: '40'
      }
    },
    result: {
      data: {
        assignment: {
          lid: '40',
          gid: 'QXNzaWdubWVudC00MA==',
          name: 'is this still workin',
          description: '<p>sadflkajsdfklajsdfasdf</p>',
          dueAt: '2018-07-11T18:59:59-06:00',
          pointsPossible: 0,
          assignmentGroup: {name: 'Assignments', __typename: 'AssignmentGroup'},
          __typename: 'Assignment'
        }
      }
    }
  }
]

beforeAll(() => {
  const found = document.getElementById('fixtures')
  if (!found) {
    const fixtures = document.createElement('div')
    fixtures.setAttribute('id', 'fixtures')
    document.body.appendChild(fixtures)
  }
  global.ENV = {}
  global.ENV.context_asset_string = 'course_1'
})

afterEach(() => {
  global.ENV = null
  ReactDOM.unmountComponentAtNode(document.getElementById('fixtures'))
})

jest.mock('timezone')

it('renders normally', async () => {
  ReactDOM.render(
    <MockedProvider mocks={mocks} removeTypename addTypename>
      <StudentView assignmentLid="40" />
    </MockedProvider>,
    document.getElementById('fixtures')
  )
  await wait(0) // wait for response
  const element = $('[data-test-id="assignments-2-student-view"]')
  expect(element).toHaveLength(1)
})
