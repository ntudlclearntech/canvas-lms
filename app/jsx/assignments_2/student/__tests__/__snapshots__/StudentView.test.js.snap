// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders normally 1`] = `
<div>
  Assignments 2 Student View
</div>
`;
